### PR TITLE
Key-based authentication system for multiplayer

### PR DESCRIPTION
**Pull request by janisozaur**
Public-private key-based authentication system for multiplayer.

**TL;DR**: this allows reliable way of **identifying clients** connecting to servers without the need for centralised authority, **persisting permissions** (actually: group assignments) and being able to recognise clients across servers (if said servers sync among themselves).

Brief overview:
This setup uses RSA asymmetric encryption algorithm to generate a pair (public and private) of keys for the client. Due to mechanics of large prime numbers, upon which RSA is based, it is very easy to *sign* a piece of information with *private key* and *verify* it with *public key*, but forging a valid signature when knowing just the *public key* and *signature* is infeasibly hard. It is a common concept in cryptography.

Upon first start with these changes, each client has a key generated for himself. When connecting to a server, a *token* is generated and sent to client for *signing*. After the client has signed the token with his private key, the signature is sent back to the server along with public key. Server then validates the signature with client's pubkey and creates a *digest* (in this case SHA1) of public key stored in a known, deterministic format (PEM). This digest is used as a human-readable ID of a given *key* (but you may think of this key as belonging to a persona) on the server, where it gets stored in a JSON file along with the group assignment.

As mentioned in the outline above, it means we can now persist group assignments, so a reconnecting client would automatically join the same group as he was assigned last time when connected.

The keys are based on username used in multiplayer, thus the client can have multiple key sets on his machine and "switch" between them using said username.

It would be easy in future to integrate this system into central authority, as users would just have to publish their public keys there and we would need an API to fetch them.

Some notes:
* This is *not* an encryption system.
* The generated key is 2048 bits long, this should be more than enough for us.
* We use OpenSSL's crypto module for handling cryptography.
* The private key is only loaded into memory when actually needed and then promptly unloaded, to reduce probability of leaking it.

Sample pubkey:
```
-----BEGIN RSA PUBLIC KEY-----
MIGJAoGBALKYuSgYuRuGVf9Zmszd7iDonmht5ruP/BJkhUQUyXQr0V8t8nOysAfb
iIl8+FCy5NaL6DjoEYnQU/eOfddVKCKr4PAy27RvC6WXmCgdXCJ4JrPbOq5Sa7kr
w9N7ifTUwtBOW6bC/7Se94BtGX9sKnpRUT9SDa4GB/C1cELELlhhAgMBAAE=
-----END RSA PUBLIC KEY-----
```

It's SHA1 sum is `45d2e18edd3ebffb01b1feee9ff9a4d4e5cebb27` and a sample `keymappings.json` file with this key looks like this:
```
[
    {
        "hash": "3081596d8324c7219e61fd46d12017a35bd6233b",
        "groupId": 3
    },
    {
        "hash": "45d2e18edd3ebffb01b1feee9ff9a4d4e5cebb27",
        "groupId": 3
    },
    {
        "hash": "662e4fec84752419c7385209abf21fe85a6d4845",
        "groupId": 0
    }
]
```

### CODE REVIEW

**Comment by janisozaur** [comment]
This does not yet solve spoofing, but this issue is easy to solve now. I would like, however, some input for that.

I was thinking about an option in the `multiplayer` window on server, where host could give their own name to a client and it would have a different colour – or perhaps some checkbox where host could tag given client as trusted.

**Comment by janisozaur** [comment]
Also: Apple, in its *infinite wisdom*, decided to drop support for OpenSSL, so they no longer provide this with system. Someone using OS X needs to fix that: either provide libs like we already do with SDL and others, or provide implementation with Apple's *bettered* crypto stuff.

**Comment by janisozaur** [comment]
Things noticed so far that I need to fix:
* [ ] crash when privkey gets corrupted
* [ ] passworded servers may not work

**Comment by Nubbie** [comment]
Not bad, my only question is, is this a 'tool' that can be worked on later on to implement other features like, a ban system?

**Comment by IntelOrca** [comment]
@Nubbie this will not provide the ability to ban, but it will help reduce the problem of griefers as permissions are remembered in a way that tackles spoofing. Servers now just need to give permissions to a user gradually as they are more trustworthy.

**Comment by janisozaur** [comment]
passworded servers should now work

File: src/network/NetworkKey.cpp [hunk 1]
```diff
@@ -333,7 +333,7 @@ std::string NetworkKey::PublicKeyHash()
     return digest_out;
 }
 
-bool NetworkKey::Sign(const char * md, const size_t len, char ** signature, size_t * out_size)
+bool NetworkKey::Sign(const uint8_t *md, const size_t len, char ** signature, size_t * out_size)
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: `uint8`?

Comment (janisozaur) [inline src/network/NetworkKey.cpp]: there's no `#include "common.h"` in here and I'd rather keep it that way

**Comment by janisozaur** [comment]
@LinusU @LRFLEW @marijnvdwerf @AaronVanGeffen can you guys look into getting OS X side of things done?

File: src/network/NetworkKey.cpp (hunk 1 above)
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: I don't want to have both `uint8_t` and `uint8` in the codebase.

**Comment by YJSoft** [comment]
OS X's OpenSSL version is `0.9.8zh`(very low!)

Possible workaround is
1. do not use evp.h at OSX
2. Force use high version of openssl at osx

**Comment by janisozaur** [comment]
@YJSoft i think all the other libs are sourced from brew, and so can be openssl: https://github.com/Homebrew/homebrew-core/blob/master/Formula/openssl.rb

**Comment by YJSoft** [comment]
@janisozaur Tried to build with Homebrew OpenSSL

**Comment by janisozaur** [comment]
@YJSoft can you provide this log in textual version, as generated by compiler? Or make a PR against my repo.

**Comment by YJSoft** [comment]
@janisozaur forgot to add new added `NetworkKey.cpp`

anyway here's build log

```
Undefined symbols for architecture i386:
  "_BIO_ctrl", referenced from:
      NetworkKey::SavePrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::PublicKeyString() in NetworkKey.o
  "_BIO_free_all", referenced from:
      NetworkKey::LoadPrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::LoadPublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::PublicKeyString() in NetworkKey.o
  "_BIO_new", referenced from:
      NetworkKey::SavePrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::PublicKeyString() in NetworkKey.o
  "_BIO_new_mem_buf", referenced from:
      NetworkKey::LoadPrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::LoadPublic(SDL_RWops*) in NetworkKey.o
  "_BIO_read", referenced from:
      NetworkKey::SavePrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::PublicKeyString() in NetworkKey.o
  "_BIO_s_mem", referenced from:
      NetworkKey::SavePrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::PublicKeyString() in NetworkKey.o
  "_CRYPTO_free", referenced from:
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.o
  "_CRYPTO_malloc", referenced from:
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.o
  "_EVP_DigestFinal", referenced from:
      NetworkKey::PublicKeyHash() in NetworkKey.o
  "_EVP_DigestInit_ex", referenced from:
      NetworkKey::PublicKeyHash() in NetworkKey.o
  "_EVP_DigestSignFinal", referenced from:
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.o
  "_EVP_DigestSignInit", referenced from:
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.o
  "_EVP_DigestUpdate", referenced from:
      NetworkKey::PublicKeyHash() in NetworkKey.o
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.o
      NetworkKey::Verify(unsigned char const*, unsigned long, char const*, unsigned long) in NetworkKey.o
  "_EVP_DigestVerifyFinal", referenced from:
      NetworkKey::Verify(unsigned char const*, unsigned long, char const*, unsigned long) in NetworkKey.o
  "_EVP_DigestVerifyInit", referenced from:
      NetworkKey::Verify(unsigned char const*, unsigned long, char const*, unsigned long) in NetworkKey.o
  "_EVP_MD_CTX_create", referenced from:
      NetworkKey::PublicKeyHash() in NetworkKey.o
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.o
      NetworkKey::Verify(unsigned char const*, unsigned long, char const*, unsigned long) in NetworkKey.o
  "_EVP_MD_CTX_destroy", referenced from:
      NetworkKey::PublicKeyHash() in NetworkKey.o
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.o
      NetworkKey::Verify(unsigned char const*, unsigned long, char const*, unsigned long) in NetworkKey.o
  "_EVP_PKEY_CTX_ctrl", referenced from:
      NetworkKey::Generate() in NetworkKey.o
  "_EVP_PKEY_CTX_free", referenced from:
      NetworkKey::~NetworkKey() in NetworkKey.o
  "_EVP_PKEY_CTX_new_id", referenced from:
      NetworkKey::NetworkKey() in NetworkKey.o
  "_EVP_PKEY_free", referenced from:
      NetworkKey::Unload() in NetworkKey.o
      NetworkKey::LoadPrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::LoadPublic(SDL_RWops*) in NetworkKey.o
  "_EVP_PKEY_get1_RSA", referenced from:
      NetworkKey::SavePrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::PublicKeyString() in NetworkKey.o
  "_EVP_PKEY_keygen", referenced from:
      NetworkKey::Generate() in NetworkKey.o
  "_EVP_PKEY_keygen_init", referenced from:
      NetworkKey::Generate() in NetworkKey.o
  "_EVP_PKEY_new", referenced from:
      NetworkKey::LoadPrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::LoadPublic(SDL_RWops*) in NetworkKey.o
  "_EVP_PKEY_set1_RSA", referenced from:
      NetworkKey::LoadPrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::LoadPublic(SDL_RWops*) in NetworkKey.o
  "_EVP_sha1", referenced from:
      NetworkKey::PublicKeyHash() in NetworkKey.o
  "_EVP_sha256", referenced from:
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.o
      NetworkKey::Verify(unsigned char const*, unsigned long, char const*, unsigned long) in NetworkKey.o
  "_OPENSSL_add_all_algorithms_noconf", referenced from:
      Network::Network() in network.o
  "_PEM_read_bio_RSAPrivateKey", referenced from:
      NetworkKey::LoadPrivate(SDL_RWops*) in NetworkKey.o
  "_PEM_read_bio_RSAPublicKey", referenced from:
      NetworkKey::LoadPublic(SDL_RWops*) in NetworkKey.o
  "_PEM_write_bio_RSAPrivateKey", referenced from:
      NetworkKey::SavePrivate(SDL_RWops*) in NetworkKey.o
  "_PEM_write_bio_RSAPublicKey", referenced from:
      NetworkKey::SavePublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::PublicKeyString() in NetworkKey.o
  "_RSA_check_key", referenced from:
      NetworkKey::LoadPrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePrivate(SDL_RWops*) in NetworkKey.o
  "_RSA_free", referenced from:
      NetworkKey::LoadPrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::LoadPublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePrivate(SDL_RWops*) in NetworkKey.o
      NetworkKey::SavePublic(SDL_RWops*) in NetworkKey.o
      NetworkKey::PublicKeyString() in NetworkKey.o
ld: symbol(s) not found for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by janisozaur** [comment]
1. you need to explicitly link with `-lcrypto` now, perhaps also `-lssl`
2. make sure you have installed openssl with `--universal` option turned on.

**Comment by YJSoft** [comment]
@janisozaur Added, It's working... I guess.(My xcode is keep returning `Command /bin/sh failed with exit code 137` so I cannot tested)

**Comment by YJSoft** [comment]
Tested with `Release` config.

It works after
1. add two new cpp,h file
2. install openssl with `--universal`
3. add `-lssl` and `-lcrypto` to linker flag

**Comment by janisozaur** [comment]
@YJSoft can you update the zip with libs at https://github.com/OpenRCT2/openrct2.github.io/tree/master/files and make a PR to my repo updating code project file?

**Comment by YJSoft** [comment]
@janisozaur OK

**Comment by LRFLEW** [comment]
> OS X's OpenSSL version is 0.9.8zh(very low!)

@YJSoft That's actually very misleading. Apple's implementation of SSL is **based** off of OpenSSL 0.9.8zh, but Apple essentially forked the project some time ago, and performs their own security updates on it. If we can get the system install of a library to work, we should always use that. It decreases the download size for the user, and it takes the burden of maintaining updates from us.

**Comment by YJSoft** [comment]
@LRFLEW I know. I'm just saying OS X's default openssl lacks `evp.h`.

**Comment by LRFLEW** [comment]
@YJSoft It actually doesn't... sort of.

I actually take back what I said. Apparently, the system install of OpenSSL has been deprecated since 10.7, and they stopped providing the headers as of the 10.11 SDK. It would work on older OS's as-is, but using our own .dylib files makes since due to the depreciation. I'll work on building them with the necessary flags.

**Comment by janisozaur** [comment]
Thanks to work of @LRFLEW and @marijnvdwerf OS X builds are now restored.

See job testing this with new libs: https://travis-ci.org/janisozaur/OpenRCT2/jobs/132249108

**Comment by janisozaur** [comment]
After a hiccup with OS X builds, which prompted to rework a large bit of Xcode project file and rebuild of libs, this is now operational again with new libs already available for downloading.

Still requires testing and figuring out way to tackle spoofing.

**Comment by IntelOrca** [comment]
* [ ] username + #N
* [ ] reload json on each access
* [ ] load the json, edit it, then save to keep custom properties

**Comment by LRFLEW** [comment]
I see why @marijnvdwerf's commits were necessary. In my sleep-deprived state, I didn't completely move all the references to the `lib` folder. Personally, I'd prefer to put them in the `lib` folder instead of the `libxc` folder, since I only created that folder because they needed to not conflict with the "source libraries", which we've long since gotten rid of. I'm ok with either solution, though.

**Comment by janisozaur** [comment]
@IntelOrca all items from your list should now be done

File: src/network/NetworkKey.cpp [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+    if (m_key != nullptr)
+    {
+        EVP_PKEY_free(m_key);
+        m_key = nullptr;
+    }
+}
+
+bool NetworkKey::Generate()
+{
+    if (m_ctx == nullptr) {
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: Brace, new line

File: src/network/NetworkKey.cpp [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+        log_error("Invalid OpenSSL context");
+        return false;
+    }
+#if KEY_TYPE == EVP_PKEY_RSA
+    if (!EVP_PKEY_CTX_set_rsa_keygen_bits(m_ctx, KEY_LENGTH_BITS)) {
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: Brace, new line

File: src/network/NetworkKey.cpp [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+        log_warning("key ok");
+    }
+    log_verbose("New key of type %d, length %d generated successfully.", KEY_TYPE, KEY_LENGTH_BITS);
+    return true;
+}
+
+bool NetworkKey::LoadPrivate(SDL_RWops * file)
+{
+    size_t size = (size_t)file->size(file);
+    if (size == (size_t)-1) {
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: Brace, new line

File: src/network/NetworkKey.cpp [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+    }
+    else if (size > 4 * 1024 * 1024)
+    {
+        log_error("Key file suspiciously large, refusing to load it");
+        return false;
+    }
+    char * pub_key = new char[size];
+    file->read(file, pub_key, 1, size);
+    BIO * bio = BIO_new_mem_buf(pub_key, size);
+    if (bio == nullptr) {
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: Brace, new line

File: src/network/NetworkKey.cpp [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+ *
+ * In particular, any of digest functions applied to a standarised key
+ * representation, like PEM, will be sufficient.
+ *
+ * @return returns a string containing key hash.
+ */
+std::string NetworkKey::PublicKeyHash()
+{
+    std::string key = PublicKeyString();
+    if (key.empty()) {
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: Brace, new line

File: src/network/NetworkKey.cpp [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+        return false;
+    }
+    if (EVP_PKEY_keygen(m_ctx, &m_key) <= 0)
+    {
+        log_error("Failed to generate new key!");
+        return false;
+    }
+    else
+    {
+        log_warning("key ok");
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: not a warning

File: src/network/network.cpp [hunk 8]
```diff
@@ -700,6 +707,40 @@ bool Network::BeginClient(const char* host, unsigned short port)
...
 
 	mode = NETWORK_MODE_CLIENT;
+	char path[MAX_PATH];
+	platform_get_user_directory(path, NULL);
+	char keyPath[MAX_PATH] = "";
+	safe_strcat(keyPath, path, MAX_PATH);
+	safe_strcat(keyPath, gConfigNetwork.player_name, MAX_PATH);
+	safe_strcat(keyPath, ".privkey", MAX_PATH);
+	if (!platform_file_exists(keyPath)) {
+		log_warning("generating key... this may take a while");
```
Comment (IntelOrca) [inline src/network/network.cpp]: not a warning

File: src/network/NetworkKey.h [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    bool SavePublic(SDL_RWops * file);
+    std::string PublicKeyString();
+    std::string PublicKeyHash();
+    void Unload();
+    bool Sign(const uint8 * md, const size_t len, char ** signature, size_t * out_size);
+    bool Verify(const uint8 * md, const size_t len, const char * sig, const size_t siglen);
+private:
+    NetworkKey ( const NetworkKey & ) = delete;
+    EVP_PKEY_CTX * m_ctx = nullptr;
+    EVP_PKEY * m_key = nullptr;
```
Comment (IntelOrca) [inline src/network/NetworkKey.h]: Convention is to just prefix with `_`, not `m_` and to align the names of the fields.

File: src/network/NetworkKey.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    bool LoadPublic(SDL_RWops * file);
+    bool SavePrivate(SDL_RWops * file);
+    bool SavePublic(SDL_RWops * file);
+    std::string PublicKeyString();
+    std::string PublicKeyHash();
+    void Unload();
+    bool Sign(const uint8 * md, const size_t len, char ** signature, size_t * out_size);
+    bool Verify(const uint8 * md, const size_t len, const char * sig, const size_t siglen);
+private:
+    NetworkKey ( const NetworkKey & ) = delete;
```
Comment (IntelOrca) [inline src/network/NetworkKey.h]: No spaces around the brackets.

File: src/network/NetworkKey.cpp [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+
+#include "NetworkKey.h"
+#include "../diagnostic.h"
+
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/pem.h>
+#include <vector>
+
+#define KEY_LENGTH_BITS 2048
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: `constexpr`

File: src/network/NetworkKey.cpp [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+#define KEY_TYPE EVP_PKEY_RSA
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: If `KEY_TYPE` is always `EVP_PKEY_RSA`, how can it not be RSA?

File: src/network/NetworkKey.cpp [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+
+NetworkKey::NetworkKey()
+{
+    m_ctx = EVP_PKEY_CTX_new_id(KEY_TYPE, NULL);
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: `nullptr`

File: src/network/NetworkKey.cpp [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+        log_error("Loaded RSA key is invalid");
+        return false;
+    }
+    BIO * bio = BIO_new(BIO_s_mem());
+    if (bio == nullptr)
+    {
+        log_error("Failed to initialise OpenSSL's BIO!");
+        return false;
+    }
+    int result = PEM_write_bio_RSAPrivateKey(bio, rsa, NULL, NULL, 0, NULL, NULL);
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: `nullptr`

File: src/network/NetworkKey.cpp [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+    *signature = nullptr;
+
+    /* Create the Message Digest Context */
+    if (!(mdctx = EVP_MD_CTX_create()))
+    {
+        log_error("Failed to create MD context");
+        return false;
+    }
+    /* Initialise the DigestSign operation - SHA-256 has been selected as the message digest function in this example */
+    if (1 != EVP_DigestSignInit(mdctx, NULL, EVP_sha256(), NULL, m_key))
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: `nullptr`

File: src/network/NetworkKey.cpp [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+    {
+        log_error("Failed to goto update digest");
+        EVP_MD_CTX_destroy(mdctx);
+        return false;
+    }
+
+    /* Finalise the DigestSign operation */
+    /* First call EVP_DigestSignFinal with a NULL sig parameter to obtain the length of the
+     * signature. Length is returned in slen */
+    if (1 != EVP_DigestSignFinal(mdctx, NULL, out_size))
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: `nullptr`

File: src/network/NetworkKey.cpp [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+    memcpy(*signature, sig, *out_size);
+    OPENSSL_free(sig);
+    EVP_MD_CTX_destroy(mdctx);
+
+    return true;
+}
+
+bool NetworkKey::Verify(const uint8 * md, const size_t len, const char * sig, const size_t siglen)
+{
+    EVP_MD_CTX * mdctx = NULL;
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: `nullptr`

File: src/network/NetworkKey.cpp [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+
+    /* Create the Message Digest Context */
+    if (!(mdctx = EVP_MD_CTX_create()))
+    {
+        log_error("Failed to create MD context");
+        return false;
+    }
+
+    if (1 != EVP_DigestVerifyInit(mdctx, NULL, EVP_sha256(), NULL, m_key))
```
Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: `nullptr`

File: src/network/network.cpp (hunk 8 above)
Comment (janisozaur) [inline src/network/network.cpp]: If there's not enough entropy available in the system, openssl may need to wait until more gets available.

File: src/network/NetworkKey.cpp (hunk 6 above)
Comment (janisozaur) [inline src/network/NetworkKey.cpp]: How about a proper tool for that instead of having to do such menial checks and changes manually? I think there's already an issue for that.

Comment (IntelOrca) [inline src/network/NetworkKey.cpp]: I don't mind if you use a tool, providing it follows the convention.

File: src/network/NetworkKey.cpp (hunk 12 above)
Comment (janisozaur) [inline src/network/NetworkKey.cpp]: This is just a safeguard, if in future someone wants to change key type to a different algorithm.

**Comment by janisozaur** [comment]
Addressed all the comments so far + rebased

File: src/network/network.h [hunk 19]
```diff
@@ -455,6 +477,9 @@ class Network
 	void Client_Handle_SHOWERROR(NetworkConnection& connection, NetworkPacket& packet);
 	void Client_Handle_GROUPLIST(NetworkConnection& connection, NetworkPacket& packet);
 	void Client_Handle_EVENT(NetworkConnection& connection, NetworkPacket& packet);
+	void Client_Handle_TOKEN(NetworkConnection& connection, NetworkPacket& packet);
+	void Server_Handle_TOKEN(NetworkConnection& connection, NetworkPacket& packet);
+	void get_key_data();
```
Comment (zsilencer) [inline src/network/network.h]: what is this get_key_data?

File: src/network/NetworkKey.cpp [hunk 20]
```diff
@@ -0,0 +1,444 @@
...
+    {
+        log_error("unknown size, refusing to load key");
+        return false;
+    }
+    else if (size > 4 * 1024 * 1024)
+    {
+        log_error("Key file suspiciously large, refusing to load it");
+        return false;
+    }
+    char * priv_key = new char[size];
```
Comment (zsilencer) [inline src/network/NetworkKey.cpp]: i notice you use new char[] and delete[] but in other areas you used std::vector for these temporary buffers

**Comment by zsilencer** [comment]
one small issue would be when the groups are changed and removed, the key group might point to the wrong one.  a solution might be to delete the keys assigned to a group when the group is removed.

File: src/network/network.h (hunk 19 above)
Comment (janisozaur) [inline src/network/network.h]: Must be some left-over, I will remove it in future commit.

File: src/network/NetworkKey.cpp (hunk 20 above)
Comment (janisozaur) [inline src/network/NetworkKey.cpp]: Yes, when the size of data is unknown beforehand, I use vector, so I can easily resize it later on.

**Comment by IntelOrca** [comment]
@zsilencer or move them to the default group, we might want to attach other data to the user and it would be a shame if those were lost.

**Comment by janisozaur** [comment]
How about `"groupId": -1`? I can make this synonymous with default group.

**Comment by LinusU** [comment]
Or maybe `"groupId": null`? Feels more explicit...

**Comment by janisozaur** [comment]
@LinusU provided jansson supports that…

**Comment by IntelOrca** [comment]
@janisozaur `json_is_null`

**Comment by IntelOrca** [comment]
Some more things:
* It doesn't save keys in a sub directory.
* If I join the game, as Player, and there is already a Player in the map, it increments the name to #2 correctly. But then if I go out and back in again I become #3.

**Comment by janisozaur** [comment]
I don't think I can (correctly) address your second point. I'm aware of its shortcomings, but it would require parsing user names and given story of https://github.com/OpenRCT2/OpenRCT2/pull/3162 I may not be the best person to do that.

Would you agree to merge this PR without this point addressed?

**Comment by IntelOrca** [comment]
@janisozaur no, but I can sort it out if you like

**Comment by IntelOrca** [comment]
Reminder for me:
* [ ] Reset player group for player if kicked

**Comment by janisozaur** [comment]
@IntelOrca I merged your PR, many thanks. I will now work on "known keys only", so something worth considering: in a dialog box to confirm kicking the user we could allow specifying *how* the host wants to kick him out:

* Simply kick, as a gentle nudge for misbehaving
* Kick + reset group assignment
* Kick + drop key from know list

**Comment by IntelOrca** [comment]
> Simply kick, as a gentle nudge for misbehaving

Well they will just join again... not a big deal for the player. Most likely the server admin wants them to not have permissions for a bit.

> Kick + reset group assignment

They can join again, but they can't retaliate. This is good because the server admin doesn't have to remember to reset their group manually to stop them from doing anything more.

> Kick + drop key from know list

Not sure what the benefit is here... other than freeing up their username to someone else. But that could confuse the admin / other players. And only works if they don't stay in the server forever.

So I think the middle one only just makes things simpler.

**Comment by janisozaur** [comment]
@IntelOrca as I mentioned earlier, I want the host to have an option to only allow players whose keys are known – this would work well with live reloading of `keymappings.json`. If now you kick a player + remove his key, he cannot even join anymore.

**Comment by IntelOrca** [comment]
@janisozaur doesn't that make passwords obsolete?

**Comment by janisozaur** [comment]
It's done now.

@IntelOrca somewhat, but I think passwords would still live on as a "simple" mechanism for casual players, while the other option would probably be used primarily by hosts like whiskey *et al.* I'm tempted to leave both options on, if one is deemed obsolete, we can simply remove it later on.

In my opinion this is more versatile than just the password, which you still cannot change on a live server.

**Comment by IntelOrca** [comment]
I think we should do some clean up now, key group map is no longer a key group map, but a user base. Of which is internally stored as a hash to user map. We can perhaps move this to a new class, RegisteredUserManager or something.

**Comment by janisozaur** [comment]
@IntelOrca https://github.com/OpenRCT2/OpenRCT2/pull/3699#issuecomment-221727923 tick that off?
